src/EditorFeatures/Core/InlineRename/InlineRenameSession.cs

```diff
@@ -664,9 +634,80 @@ public void Cancel()
 
-            LogRenameSession(RenameLogMessage.UserActionOutcome.Canceled, previewChanges: false);
+            DismissUIAndRollbackEditsAndEndRenameSession(RenameLogMessage.UserActionOutcome.Canceled, previewChanges: false);
+        }
+
+        private void DismissUIAndRollbackEditsAndEndRenameSession(
+            RenameLogMessage.UserActionOutcome outcome,
+            bool previewChanges,
+            Action finalCommitAction = null)
+        {
+            // Note: this entire sequence of steps is not cancellable.  We must perform it all to get back to a correct
+            // state for all the editors the user is interacting with.
+
+            // Remove all our adornments and restore all buffer texts to their initial state.
             DismissUIAndRollbackEdits();
-            EndRenameSession();
+
+            _triggerView.Caret.PositionChanged += LogPositionChanged;
+
+            // We're about to perform the final commit action.  No need to do any of our BG work to find-refs or compute conflicts.
+            _cancellationTokenSource.Cancel();
+            _conflictResolutionTaskCancellationSource.Cancel();
+
+            // Perform the actual commit step if we've been asked to.
+            finalCommitAction?.Invoke();
+
+            // Log the result so we know how well rename is going in practice.
+            LogRenameSession(outcome, previewChanges);
+
+            // Remove all our rename trackers from the text buffer properties.
+            RenameTrackingDismisser.DismissRenameTracking(_workspace, _workspace.GetOpenDocumentIds());
+
+            // Log how long the full rename took.
+            _inlineRenameSessionDurationLogBlock.Dispose();
+
+            _triggerView.Caret.PositionChanged -= LogPositionChanged;
+
+            return;
+
+            void DismissUIAndRollbackEdits()
+            {
+                _dismissed = true;
+                _workspace.WorkspaceChanged -= OnWorkspaceChanged;
+                _textBufferAssociatedViewService.SubjectBuffersConnected -= OnSubjectBuffersConnected;
+
+                // Reenable completion now that the inline rename session is done
+                _completionDisabledToken.Dispose();
+
+                foreach (var textBuffer in _openTextBuffers.Keys)
+                {
+                    var document = textBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+                    var isClosed = document == null;
+
+                    var openBuffer = _openTextBuffers[textBuffer];
+                    openBuffer.DisconnectAndRollbackEdits(isClosed);
+                }
+
+                this.UndoManager.Disconnect();
+
+                if (_triggerView != null && !_triggerView.IsClosed)
+                {
+                    _triggerView.Selection.Clear();
+                }
+
+                RenameService.ActiveSession = null;
+            }
+
+            void LogPositionChanged(object sender, CaretPositionChangedEventArgs e)
```

Comment:
We can get rid of the `LogPositionChanged` portion now. It is being resolved

Reply:
Oh?  Do tell, I'm interested :-)